proposals/2045-column-lineage-endpoint.md

```diff
@@ -0,0 +1,143 @@
+
+- OpenLineage defines a [column-level lineage facet]- (https://github.com/OpenLineage/OpenLineage/blob/ff0d87d30ed6c9fe39472788948266a6d3190585/spec/facets/ColumnLineageDatasetFacet.md). 
+- Marquez has a lineage endpoint `GET /api/v1/lineage` that returns the current lineage graph connected to a job or a dataset
+
+### New Elements
+We propose to add the following:
+- Add column lineage to the lineage endpoint
+- A new column-lineage endpoint leveraging the column lineage facet to retrieve lineage for a given column.
+- Point-in-time upstream (dataset or column level) lineage given a version of a dataset.
+
+## Proposal
+
+### add column lineage to existing endpoint
+In the GET /lineage api, add column lineage to DATASET nodes' data
+```diff
+{
+  "id": "dataset:food_delivery:public.categories",
+  "type": "DATASET",
+  "data": {
+    "type": "DATASET",
+    "id": {
+      "namespace": "food_delivery",
+      "name": "public.categories"
+    },
+    "type": "DB_TABLE",
+    ...
+    "fields": [{
+		...
+    }],
+>   columnLineage: {
+>     "a": {
+>       inputFields: [
+>          {namespace: "ns", name: "name", "field": "a"},
+>          ... other inputs
+>       ],
+>       transformationDescription: "identical",
+>       transformationType: "IDENTITY"
+>     },
+>     "b": ... other output fields
+>   }
+  },
+  "inEdges": [{
+    "origin": "job:food_delivery:etl_orders.etl_categories",
+    "destination": "dataset:food_delivery:public.categories"
+  }],
+  "outEdges": [{
+    "origin": "dataset:food_delivery:public.categories",
+    "destination": "job:food_delivery:etl_orders.etl_orders_7_days"
+  }]
+}
+```
+
+### add a column-level-lineage endpoint:
+
+```
+GET /column-lineage?nodeId=dataset:food_delivery:public.delivery_7_days&column=a
+```
+`column` is a ne parameter that must be a column in the schema of the provided dataset `nodeId`.
+
+The logic is layered on the existing lineage endpoint, filtering down to the datasets that contribute to that column.
+It only returns dataset nodes.
+
+```diff
+{
+  graph: [
+    {
+       "id": "dataset:db1:table2",
+       "type": "DATASET",
+       data: {
+         namespace: "DB1",
+         name: "table2",
+>       columnLineage: {
+>           "a": {
+>              inputFields: [
+>                {namespace: "DB1", name: "table1, "field": "a"}
+>              ],
+>              transformationDescription: "identical",
+>              transformationType: "IDENTITY"
+>           },
+>          "b": ... other output fields
+>        }
+      },
+      ...
+  }
+  ]
+}  
+```
+
+### Point in time upstream lineage
+return historical upstream lineage from a given Dataset version. 
+This adds the version element to the nodeId in both the existing `/api/v1/lineage` and newly proposed `/api/v1/column-lineage` endpoint 
+```
+GET /lineage?nodeId=dataset:food_delivery:public.delivery_7_days:{version}
+GET /column-lineage?nodeId=dataset:food_delivery:public.delivery_7_days:{version}&column=a
```

Comment:
Couldn't we allow providing a timestamp within URL instead of dataset version? 
```
GET /lineage?nodeId=dataset:food_delivery:public.delivery_7_days&timestamp=1661417230
```
Dataset version is an internal Marquez label and being able to retrieve information for yesterday, two days ago, etc. seems to me straightforward for the user.